xds/internal/balancer/priority/balancer.go

```diff
@@ -0,0 +1,221 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package priority
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/internal/buffer"
+	"google.golang.org/grpc/internal/grpclog"
+	"google.golang.org/grpc/internal/hierarchy"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/xds/internal/balancer/balancergroup"
+)
+
+const priorityBalancerName = "priority_experimental"
+
+func init() {
+	balancer.Register(&priorityBB{})
+}
+
+type priorityBB struct{}
+
+func (pbb *priorityBB) Build(cc balancer.ClientConn, _ balancer.BuildOptions) balancer.Balancer {
+	return newPriorityBalancer(cc)
+}
+
+func (pbb *priorityBB) Name() string {
+	return priorityBalancerName
+}
+
+type priorityBalancer struct {
+	logger           *grpclog.PrefixLogger
+	cc               balancer.ClientConn
+	bg               *balancergroup.BalancerGroup
```

Comment:
Why do we need a balancerGroup here?

Reply:
A child policy is started by being added to the balancer group.

1. p0 is added to balancer group
1. if p0 is ready, use it; otherwise, add p1 to balancer group

Balancer group is responsible for forwarding the updates, and caching (when a child is removed).